Connection options

richard-walters opened the pull request:
Add support to provide database user password and configure SSL mode for connection.

gajohnson commented:
Thanks @richard-walters I'll test & review some time this week. I am a little concerned about the addition of a password command line arg as that should only really be used in an introductory example. Database credentials can be unintentionally exposed by doing it that way. The intention for utilities like this would be to use the pgpass file to safely store database credentials. https://www.postgresql.org/docs/9.6/static/libpq-pgpass.html

At the very least, I would expect this change to have no impact on users that are relying on a pgpass file for creds. The way this is implemented will default to empty passwords so you should check that psycopg handles this correctly if there is a pgpass file or environment variable.

gajohnson commented:
@richard-walters Did you get a chance to look at how this behaves with a .pgpass file?

richard-walters commented:
Hi @gajohnson,

Sorry for not getting back to you sooner. I have to admit I'm quite new to postgres so didn't realise the password could be set in a pgpass file. I set up pg2kinesis in a docker container that retrieves secrets from credstash at runtime, so passing the password as a command line arg in my use case didn't seem too much of a concern. I've switched my implementation to create a pgpass file instead, so I'm happy to remove the password command line option from the pull request.

The implementation did work fine with a pgpass file when not providing the command line arg, and then specifying the arg overwrote the password in the pgpass file. The implementation was also fine when the user had no password set.

Do you still want the other changes in the pull request? It turns out after testing again that I don't need the sslmode arg either, but it seems sensible to provide this option.

I do have a few questions for you regarding the pg2kinesis. Firstly, I've set up the utility as a solution for capturing database events in S3. I have a Firehose delivery stream pushing the events unmodified to a bucket in S3, and then notifying a lambda to transform these as Avro and segment them so that we can run queries over the data using Amazon Athena. Firehose concatenates the events into a single string but doesn't provide an option for delimiting records as it does so. In my fork of pg2kinesis I add \n to the end of each event as a record separator so that the S3 file is easier to parse. Would you be happy for me to submit a second pull request to optionally set record separators when running the utility?

Secondly, how do you handle the 'at least once' nature of the event stream? I've been following this blog post as a guide when setting up our pipeline, https://www.simple.com/engineering/a-change-data-capture-pipeline-from-postgresql-to-kafka, and they capture the lsn in order to de-duplicate events downstream.

Finally, and most importantly, we've noticed that the free disk space on our RDS instance decreases linearly with pg2kinesis in place when there are no writes to our tables (it actually brought our dev instance down at the weekend as all the disk space was consumed!). As soon as an event is processed by pg2kinesis, the replication lag reduces to zero as expected as the disk space taken by the previously unacknowledged WAL is recovered. The write IOPS monitor in AWS shows that there is write activity occurring every 5 minutes, which I presume is caused by an internal postgres process such as stats collection or checkpointing. It appears that it is this activity that is causing the replication lag to increase linearly and consume disk space. Have you come across this before, and is there a way pg2kinesis could handle this?

Regards,
Richard.

gajohnson commented:
Thanks @richard-walters!

I'm glad the .pgpass file got you unstuck. If we are going to add a command like arg for passwords I would rather make that flag a boolean to indicate that you want to provide a password interactively rather than specify it right at the command line. The difference is that by providing the password in the arguments you will expose that password to anyone else on the machine who can list running processes. This could be a serious breach in the case of a production database password. This is the pattern that `psql` uses so it would be familiar to other users of pg2kinesis. You can read more at https://www.postgresql.org/docs/9.5/static/app-psql.html

> --password / Force psql to prompt for a password before connecting to a database.

Also note that if we are prompting for a password we should be using the python built-in getpass.

For the newline issues in firehose, I also noticed that. I suspect that has something to do with the aws_kinesis_agg package we are using to batch records together. More investigation will be needed before we can say what the correct fix is because we do not have that issue when consuming from kinesis with `amazon_kclpy_helper.py`. We can open an issue for that now.

The at-least-once property of the event stream comes from the fact that the logical replication slot is durable across db restarts, and that we peek changes from the slot and wait for an OK from kinesis before advancing. If pg2kinesis restarts in between a put to kinesis and advancing the slot you will see duplicate records. Each record has the corresponding transaction id on it which will help with deduplication on the consumer. Agreed that we can use lsn to be more granular with that.

For your disk space problem: if you have a logical replication slot open but you are not consuming from it, your disk will fill up. This will also be the case if you have long-running or stuck transactions. By long-running I mean a transaction open for the entire duration of the disk consumption you saw. So long as pg2kinesis is running successfully (and you only have 1 slot open) disk space should not be an issue. Can you confirm that when you saw disk fill on your database that 1) pg2kinesis was reading from the slot, 2) you only have the one slot open, and 3) you don't have transactions open that do not close?

richard-walters commented:
I've reverted the password option commit. Let me know if you still want to merge in the other changes.

I don't think the delimiting issue is because of an issue with pg2kinesis. Originally, I was using a lambda to do a data transformation in firehose before dumping the data in S3. The lambda picked up each batch of events from the kinesis stream as expected, but firehose then concatenated the transformed events into a single string when delivering to S3. It turns out its up to you to provide a record separator when using firehose. I added a newline character to each event in the lambda, but later decided the lambda transformation was unnecessary so I had to drop the newline character on each event in pg2kinesis instead.

I guess either users of pg2kinesis will be required to use a lambda with firehose to overcome this issue, or pg2kinesis provides an option to append a record separator to each event. What do you think?

I was being a bit silly about de-duplication; I wasn't capturing the transaction ID when transforming the pg2kinesis events so forgot it was there!

For the disk space issue, pg2kinesis was definitely reading the slot, and only one slot was created. I'll check that there's no open transactions. However, as I say, I think some internal process, which causes small spikes in our write IOPS every 5 minutes, is advancing the LSN causing the DB to think that our slot is lagging, when actually no activity that could be read by pg2kinesis has occurred. We're observing the same behaviour as this post in the AWS forums: https://forums.aws.amazon.com/thread.jspa?messageID=780025. There's nothing special about our RDS instance, it's a vanilla postgres 9.6.5 but with logical_replication enabled. I'll investigate further and create a demo repo in terraform to share with you.

I really appreciate your help with this! pg2kinesis has allowed us to quickly bootstrap a CDC pipeline to power our reporting requirements, so thanks for creating the utility.
